autoload/SimplenoteUtilities.py

```diff
@@ -327,51 +329,159 @@ def display_note_in_scratch_buffer(self, note_id=None):
                 vim.command("resize " + vim.eval("s:listsize"))
                 vim.command("wincmd p")
 
+    def pre_rename_buffer(self):
+        """ the only way to detect if user is saving with :saveas command.
+            this function is executed before the actual saving """
+        vim.command("let s:renaming = 1") # we just need to know if user is renaming this buffer
+
+
     def update_note_from_current_buffer(self):
         """ updates the currently displayed note to the web service or creates new """
-        note_id = self.get_current_note()
-        content = "\n".join(str(line) for line in vim.current.buffer[:])
-        # Need to get note details first to assess remote markdown status
-        note, status = self.simplenote.get_note(note_id)
-        if status == 0:
-            if (vim.eval("&filetype") == "markdown"):
-                if "systemtags" in note:
-                    if ("markdown" not in note["systemtags"]):
-                        note["systemtags"].append("markdown")
-                else:
-                    note["systemtags"] = ["markdown"]
-            else:
-                if "systemtags" in note:
-                    if ("markdown" in note["systemtags"]):
-                        note["systemtags"].remove("markdown")
-            # To merge in we need to send version.
-            note, status = self.simplenote.update_note({"content": content,
-                                                      "key": note_id,
-                                                      "version": self.note_version[note_id],
-                                                      "systemtags": note["systemtags"]})
+
+        # what information do we have?
+        amatch=vim.eval('expand("<amatch>")') # the desired path+file name (:w command parameter, for example)
+        currentfile=vim.eval("expand('%:p')") # the filename of this buffer
+        renaming=vim.eval("s:renaming")       # is the user changing THIS buffer's file name? (:saveas)
+
+        # based on the available information, let's try to find out what the
+        # user action is
+        if os.path.basename(currentfile) != os.path.basename(amatch):
+            userAction="writingBufferToADifferentFile" # user is executing :w <newfile>
+        elif renaming == "0":
+            userAction="updatingNote"
+        else:
+            userAction="renamingBuffer"
+
+        if userAction=="updatingNote":
```

Comment:
This is a really big `if` statement. I bet it would be easier to maintain if it were extracted into 3 separate methods (`update_note()`, etc.)